### PR TITLE
switch to OAuth 2

### LINKED ISSUE #7: switch to using oauth2 as the authtoken is being deprecated
https://www.site24x7.com/help/api/index.html#authentication

which states: 

```New authtokens will be deprecated by 31st August, 2019 and the existing ones will be deprecated by 30th November, 2019. Please migrate to OAuth for a more secure access.```

### PR DESCRIPTION
**Pull request by uwedeportivo**
Resolves https://github.com/sourcegraph/terraform-provider-site24x7/issues/7

Uses sample code from 

https://zohodiscussions.com/viewFile.do?fileId=18151000001568467&forumGroupURL=site24x7

(heavily modified, cleaned up and simplified)

### CODE REVIEW

**Comment by uwedeportivo** [comment]
i will add tests over the weekend and test it myself by hand

File: main.go [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+	"flag"
+	"log"
+
 	"github.com/hashicorp/terraform/plugin"
+	"github.com/hashicorp/terraform/terraform"
 	"github.com/sourcegraph/terraform-provider-site24x7/site24x7"
+	"github.com/sourcegraph/terraform-provider-site24x7/site24x7/oauth"
 )
 
+var oauthFile = flag.String("oauth_file", "", "path to the oauth_file")
```
Comment (unknwon) [inline main.go]: I don't know... just that using underscore in flags seems weird in Go:

```suggestion
var oauthFile = flag.String("oauth-file", "", "path to the OAuth file")
```

File: site24x7/oauth/oauth.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+)
+
+func setRequestHeaders(request *http.Request, accessToken string) {
+	request.Header.Set("Content-Type", "application/json;charset=UTF-8")
+	request.Header.Set("Accept", "application/json; version=2.0")
+	request.Header.Set("Authorization", "Zoho-oauthtoken " + accessToken)
+}
+
+func getUrl(urlValues url.Values) string {
+	baseAccurl := fmt.Sprintf("%s%s", tokenDomain, tokenApi)
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: nit:
```suggestion
	baseAccURL := fmt.Sprintf("%s%s", tokenDomain, tokenApi)
```

File: site24x7/oauth/oauth.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	timeout     = time.Second
+)
+
+func setRequestHeaders(request *http.Request, accessToken string) {
+	request.Header.Set("Content-Type", "application/json;charset=UTF-8")
+	request.Header.Set("Accept", "application/json; version=2.0")
+	request.Header.Set("Authorization", "Zoho-oauthtoken " + accessToken)
+}
+
+func getUrl(urlValues url.Values) string {
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: nit:
```suggestion
func getURL(urlValues url.Values) string {
```

File: main.go [hunk 4]
```diff
@@ -1,12 +1,28 @@
...
+
 func main() {
+	flag.Parse()
+
+	ator, err := oauth.NewAuthenticator(*oauthFile)
```
Comment (unknwon) [inline main.go]: `ator` feels like made by some non-native speaker (i.e. me) would come up with 😂 I used `iset` for `intersection` in another PR...

File: site24x7/oauth/oauth.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+type Authenticator struct {
+	// Protects tkns.AccessToken (updated in refresh goroutine)
+	sync.Mutex
+	tkns      tokens
+	storePath string
+}
+
+func (ator *Authenticator) getClient() *http.Client {
+	trTls11 := &http.Transport{
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: Should it be `trTls13` (actually uses `tls.VersionTLS13`)?

nit:

 ```suggestion
	trTLS11 := &http.Transport{
```

File: site24x7/oauth/oauth.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+		return err
+	}
+	defer res.Body.Close()
+
+	err = json.NewDecoder(res.Body).Decode(&result)
+	if err != nil {
+		return err
+	}
+	ator.Lock()
+	ator.tkns.AccessToken = result["access_token"].(string)
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: Is it guaranteed `result["access_token"]` always have a string type value at this point?

Comment (unknwon) [reply site24x7/oauth/oauth.go]: Based on line 84, I don' think so?

Using this seems safer with type assertion:
```suggestion
	ator.tkns.AccessToken, _ = result["access_token"].(string)
```

File: site24x7/oauth/oauth.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	ator.Unlock()
+
+	ator.tkns.ExpiresInSec = result["expires_in_sec"].(float64)
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]:
```suggestion
	ator.tkns.ExpiresInSec, _ = result["expires_in_sec"].(float64)
```

File: site24x7/oauth/oauth.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+			if err != nil {
+				fmt.Printf("failed to refresh access token: %b", err)
+			}
+		}
+	}()
+}
+
+// AccessToken returns the access token.
+func (ator *Authenticator) AccessToken() string {
+	ator.Lock()
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: What about using [`sync.RWMutex`](https://pkg.go.dev/sync?tab=doc#RWMutex)?

File: site24x7/oauth/oauth.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	defer ator.Unlock()
+
+	return ator.tkns.AccessToken
+}
+
+// NewAuthenticator creates an authenticator that will acquire an access token by reading it from the specified
+// file if it exists or by requuesting it from Zoho and saving it to the specified file. It also schedules a
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]:
```suggestion
// file if it exists or by requesting it from Zoho and saving it to the specified file. It also schedules a
```

File: main.go (hunk 4 above)
Comment (uwedeportivo) [reply main.go]: it's short but still distinct

File: site24x7/oauth/oauth.go (hunk 5 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: no clue, i got this stuff from the sample code

File: site24x7/oauth/oauth.go (hunk 8 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: it doesn't really matter :-)

Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: done

File: site24x7/oauth/oauth.go (hunk 7 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: done

File: site24x7/oauth/oauth.go (hunk 6 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: done

File: site24x7/oauth/oauth.go (hunk 3 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: done

File: main.go (hunk 1 above)
Comment (uwedeportivo) [reply main.go]: done

**Review by uwedeportivo — COMMENTED** [review]
.

File: site24x7/oauth/oauth.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+		ator.tkns.ExpiresInSec = 300
+		return err
+	}
+	return nil
+}
+
+func (ator *Authenticator) scheduleRefresh() {
+	go func() {
+		for {
+			timer := time.NewTimer(time.Second * time.Duration(math.Max(ator.tkns.ExpiresInSec - 10, 2)))
```
Comment (unknwon) [inline site24x7/oauth/oauth.go]: The calculation of `math.Max(ator.tkns.ExpiresInSec - 10, 2)` looks strange... what's the definition of its result?

**Review by unknwon — APPROVED** [review]

File: site24x7/oauth/oauth.go (hunk 10 above)
Comment (uwedeportivo) [reply site24x7/oauth/oauth.go]: it's 10 seconds less than ExpiresInSec unless  that is smaller than 2 at which point it's 2

**Comment by uwedeportivo** [comment]
I made a small utility to generate the oauth JSON file.

Build both `main.go` and `site24x7/oauth/cmd/main.go` and set up file `main.tf`
```
resource "site24x7_website_monitor" "my-monitor" {
  display_name="foo"
  website="foo.com"
}
```

Then:

- launch `site24x7/oauth/cmd/site24x7-oauth` to create file `~/.site24x7-oauth.json` (or give it an argument to a different location)
- visit https://api-console.zoho.com/client to create a self-client and copy client id, client secret and generate code into `~/.site24x7-oauth.json`
- launch `site24x7/oauth/cmd/site24x7-oauth` again and it will acquire an access token. if it succeeded it will save the necessary tokens into the JSON file and print out the env var to set
- execute `SITE24X7_AUTHTOKEN_FILE=~/.site24x7-oauth.json terraform init`
- execute `SITE24X7_AUTHTOKEN_FILE=~/.site24x7-oauth.json terraform plan`
... and so on (terraform commands)

**Review by keegancsmith — APPROVED** [review]
LGTM, didn't review too closely though. Before merging, can you create a PR in our infrastructure repo and test that our site24x7 terraform files don't break.

**Comment by uwedeportivo** [comment]
i tested on sourcegraph/infrastructure and it worked as advertised

**Pull request merged**